Home_work22/DynamicArray.h

```diff
@@ -0,0 +1,134 @@
+#pragma once
+#include <iostream>
+template <typename T> class DynamicArray {
+private:
+    std::size_t size = 0;
+    std::size_t capacity = 0;
+    T* data;
+public:
+
+    //default c-tor
+    DynamicArray() {
+        data = new T[size];
+    };
+
+    //c-tor with size
+    DynamicArray(std::size_t newCapacity) {
+        capacity = newCapacity;
+        data = new T[newCapacity];
+    };
+
+    //copy c-tor
+    DynamicArray(const DynamicArray& other) {
+        this->size = other.size;
+        capacity = other.capacity;
+        data = new T[size];
+        for (int i = 0; i < other.size; ++i) {
+            data[i] = other.data[i];
+        }
+    };
+
+    //free allocated memory
+    ~DynamicArray() {
+        delete[] data;
+    };
+
+    DynamicArray& operator=(const DynamicArray& other) {
+        this->size = other.size;
+        this->capacity = other.capacity;
+        delete[] data;
+        data = new T[size];
+        for (int i = 0; i < size; ++i) {
+            data[i] = other.data[i];
+        }
+        return *this;
+    };
+
+    T& operator[](std::size_t index) {
+        return data[index];
+    };
+
+    bool operator==(DynamicArray& other) {
+        if (size == other.size) {
+            for (int i = 0; i < size; ++i) {
+                if (data[i] != other.data[i]) return false;
+            }
+            return true;
+        }
+        else return false;
+    }
+    //delete previous memory, create new inner array with updated size
+    void setSize(std::size_t newSize) {
+        delete[] data;
+        size = newSize;
+        capacity = size;
+        data = new T[size];
+    };
+    std::size_t getSize() const {
+        return this->size;
+    };
+
+    //Clear internal memory, set size to 0
+    void clear() {
+        delete[] data;
+        size = 0;
+        data = new T[capacity];
+    };
+
+
+    // Allocate new memory for new element
+    // Copy old content to new inner array
+    // insert element at the last index
+    // update size
+    void pushBack(T element) {
+        if (capacity > size) {
+            data[size] = element;
+            size += 1;
+        }
+        else {
+            T* newArray = new T[size + 1];
```

Comment:
capacity should be incremented as well.  Possible approach is: capacity = size + size*1/2